mitodl/open-discussions
# Reorganize frontend(s) to use yarn workspaces for separation

#### Pre-Flight checklist


#### What are the relevant tickets?
Related to https://github.com/mitodl/hq/discussions/95

#### What's this PR do?
A few related things:

1. Reorganizes the frontend code to use yarn workspaces, similar to https://github.com/mitodl/mitxonline/pull/514. This should let develop independent frontends (with shared dependencies, if needed).
2. Update django-webpack-loader to 1.4, and remove our customized stuff. The custom stuff was, I believe, for overwriting publicPath. We can do this with django-webpack-loader, and the upgraded version of django-webpack-loader supports multiple webpack configs, whcih we needed.
3. Updates to Node 16
4. Updates several npm dependencies. Mostly these were minor version bumps. A noteable exception is webpack, see comments in PR.
5. Adds a demo frontend "infinite-corridor".
There are almost no changes to application code.

#### How should this be manually tested?

Test in dev mode:

1. Ensure your .env file includes
    ```
    DEBUG=True
    NODE_ENV=development
	OPEN_DISCUSSIONS_HOSTNAME // hostname, e.g., yourname.od.odl.local ... defaults to od.odl.local
    ```
2. Rebuild docker to get new python dependency (updated version of django-webpack-loader): `docker compose up --build`
3. Visit http://yourname.od.odl.local:8063 ... you should see the MIT Open app. Try making a change in react code (e.g., frontends/open-discussions/src/components/IntroCard.js). The app should reload.
4. Test anything else you want in the frontend.
5. Visit http://yourname.od.odl.local:8063/infinite ... you should see a placeholder infinite corridor frontend. Try making changes in frontends/infinite-corridor/src/entry/root.tsx ... The page should reload.

Test in prod mode:

1. Ensure your .env file includes
    ```
    DEBUG=False
    NODE_ENV=production
	OPEN_DISCUSSIONS_HOSTNAME // hostname, e.g., yourname.od.odl.local ... defaults to od.odl.local
    ```
2. Restart docker.
3. View the json files in `webpack-stats/` and ensure the publicpaths point to `static/`.
4. View http://yourname.od.odl.local:8063 and http://yourname.od.odl.local:8063/infinite ... The apps should work but will not live reload.


#### Where should the reviewer start?
⚠️ ⚠️ Strongly suggest reviewing https://github.com/mitodl/open-discussions/pull/3593/commits/c6ea8fd5117b4e6622045fb8dbdd702053dad500) ... it contains all the changes other than moving code / creating new placeholder frontend.

#### Any background context you want to provide? 
See github comments (most easily viewable in https://github.com/mitodl/open-discussions/pull/3593/commits/c6ea8fd5117b4e6622045fb8dbdd702053dad500)
